### ISSUE TITLE
Discussion on the project's Stack. 

### ISSUE THREAD

**Issue by VernL** [RichardLitt/the-travel-shelf#23]
Hey all, let's use everything we've learned to discuss our options and choose a Stack!

**Comment by VernL** [comment]
Personally, I feel that as beginners, we should stick to mature languages and frameworks with strong communities, good documentation and plenty of learning resources. Also, learning technologies that are popular within our community may improve our ability to participate and contribute to future projects. 

Let’s begin by choosing a language for our server-side code. 

I came across [this]( http://githut.info/) page, which ranks each language based on number of active repositories and total pushes. (spolier alerts, Javacript is #1). Although popularity alone may not cover all strengths and weaknesses of a language, I feel it’s a strong indicator.  

The fact that ruby has declined from # 2 in Q2/12 to # 6 in Q4/14 is enough to turn me away. 

Go and Rust are so far down the list that I don’t feel like they would be a good choice for our first project. 

This leaves Python/Django, Java, and JavaScript/Node.js in the race. 

These are all very popular options and are probably all great choices. However, my choice is Node.js and it comes down to one thing: **JavaScript**.  As explained in my response to #11 , using JavaScript seems like a great choice for our team for a verity of reason including: 
- Giving us a chance to apply what we are learning in the JS workshops. 
- Avoid the hassle of us all learning a new language, leading to the next point,
- Facilitate collaboration within our team as everyone will be able to understand both front-end and back-end code. 

Now onto front end!

For the same reasons as above, for many of us this being our first project, I think it would be a safe choice to keep our heads down and follow the crowd.

I came across this page https://github.com/showcases/front-end-javascript-frameworks which ranks front-end JavaScript frameworks based on how many stars they have received on GitHub. Other interesting metrics may be number of contributors and # of forks, summarized in the table below:

| Framework | Stars | Forks | Contributors | 
|--------------|------|-------|----------------|
| Facebook / React           | 76,041| 14,416 | 1,060 |
| Vuejs / vue | 67,659 | 9,754 |  135 | 
| Angular / Angular.JS | 57,138 | 28,491|1,604|
|Emberjs / ember.js |18,258 | 3,750 |680|

We see that React, vuejs and AngularJS are all very popular. Interestingly, vuejs has significantly less contributors than everything else.  This alone is of course not enough info to make a decision...

Angular JS is actually outdated and has been replaced by angular 2. Angular 2 uses TypeScript not JS. Using the same logic as for the back-end language, learning TypeScript may not be ideal for our team at this time...However, lots of people love TypeScript (including my brother :) 

I personally have already learned vue, and find it a pleasure to use, however was disappointed when I realized that nobody else is using it. My strongest desire is to learn skills, languages and technologies that will allow me to contribute within the community,  so I would sadly not recommend choosing vuejs. 

Even though it's supposed to be hard to learn, I know at least one other team is using React, so it could be a good choice as it could facilitate collaboration. Aslo, it is possible to use React components without building a full fledged single page applications. (not possible with Angular 2)

I found an interesting [video](https://www.youtube.com/watch?v=KMX1mFEmM3E) comparing Angular vs React vs Vue.js (15 minutes) 

This is getting pretty long winded so I'll leave the floor open for somebody else!

**Comment by Jojokats** [comment]
I agree that as beginners, we should focus on languages and frameworks with strong communities, good documentation and plenty of learning resources. 

Since we will be focusing on javaScript & React at the Workshops it’s to our advantage to focus on these languages and frameworks for our project.

For backend,  I’m debating between Python & Node.js. I read that Python is an object-orientated language that closely resembles the English language which makes it a great language to learn for beginners and Lab12 is also providing Python workshops.

Also read that it's the best language as of 2017 to learn first. [Have a look at the list here.](https://www.slant.co/topics/25/~best-programming-language-to-learn-first)

**Comment by Jojokats** [comment]
I looked into Airbnb to see what programming languages they are using and they use:
+ ReactJS 
+ Ruby 
as per [Packt](https://www.packtpub.com/books/content/what-software-stack-does-airbnb-use)

Ruby on Rails is a framework that could be interesting to work with as well and they're job demands for this programming languages.

**Comment by VernL** [comment]
Several items in #8 turn me away from ruby including:
 - Poor documentation, speed, & declining popularity. 

I watched part of the Rails tour featured on [the Rails homepage](http://rubyonrails.org/). It kind of looks like magic, in the sense that a lot of code is "created" out of thin air. 

Thoughts?

**Comment by HanieJ** [comment]
I agree with @VernL (JS/React)! I think to master a full-stack even with some limited tools takes lots of time. I love to learn Python, but I would rather focus on JS/Node(with db) and React(maybe Sass/a bit of sketch too). This is a full dish! If we were able to eat it all, we can start learning Python as well!

**Comment by RichardLitt** [comment]
> I read that Python is an object-orientated language that closely resembles the English language which makes it a great language to learn for beginners and Lab12 is also providing Python workshops.

Python doesn't necessarily demand you follow object-orientated code. And I don't think it resembles the English language enough to really be a help. Having workshops would help, but we also have those for Node. 

I think that we shouldn't go with Ruby, for the reasons that Vernon pointed out. It would be interesting to do for job stuff - a lot of people use it - but I think we can also get that from Node. 

What I was hoping others would mention in the language considerations was the ecosystem as a whole. Learning a language is as much about learning what are the hundreds of little tools for that language as it is learning the syntax or conventions. I don't know Python or Ruby's ecosystem nearly as well as Node, which is discouraging. For a language like Go, that'd be OK - it'd be fairly trivial to learn what others are using, since there is so little out there. But that would mean we'd have to work with little to no documentation. With Ruby or Python, it'd be very hard to google to find the right solution to a problem, as there are dozens of solutions. We'd have to search a lot to find the best ones. With JavaScript, I think we can narrow in a bit further because of my experience. I'm not entirely sure about this, though, and don't want us to rely on it. 

I think we should go with Node for this reason. 

> Angular 2 uses TypeScript not JS. 

Let's abandon using Angular. I think that TypeScript and other compile-to-JS languages lead to fragmentation of the community, which is not good. I'm not a fan of them. 

> My strongest desire is to learn skills, languages and technologies that will allow me to contribute within the community, so I would sadly not recommend choosing vuejs.

I agree with this.

**Comment by RichardLitt** [comment]
I went through the issues for all of the other languages and projects and closed them. Let me know if anyone has any issues on any of them, and we can discuss them on those threads! For now, it is clear to me that NodeJS and React are probably the best options for us. Does it seem that way to you? Do you feel like you know more about the different languages and options?

**Comment by HanieJ** [comment]
Guys, Did you see the news that Corin shared? 
https://thenextweb.com/dd/2017/09/19/should-developers-be-afraid-of-zuckerbergs-bearing-gifts/

**Comment by VernL** [comment]
I agree and am happy we had the discussion. 
TIL there are a tone of cool/fun/interesting... options, but before venturing into the wild and getting lost in a sea of possibilities, it may be beneficial to begin by learning the tools that are actively being used within the community/job market.   
Cool exercise, cheers.

**Comment by VernL** [comment]
> Guys, Did you see the news that Corin shared?
https://thenextweb.com/dd/2017/09/19/should-developers-be-afraid-of-zuckerbergs-bearing-gifts/

I personally think that when React dies, everyone will just move onto the next thing. I disliked facebook before the article and still dislike them now :P

**Comment by Jojokats** [comment]
LGTM, I'm all for NodeJS & React.

**Comment by RichardLitt** [comment]
> Guys, Did you see the news that Corin shared?
thenextweb.com/dd/2017/09/19/should-developers-be-afraid-of-zuckerbergs-bearing-gifts

I did. I'd rather not use React for that reason - but we're going to run into that issue, most places. This is one of the reasons I elevated choo. It's under the [MIT](https://github.com/choojs/choo/blob/master/LICENSE) license, and there is no patents issue. It'll also probably afford us the possibility of giving back to it, which is something I want to be sure we can do - give some code back to the community who we borrow code from.
